### PR TITLE
Node: inbound stats & test

### PR DESCRIPTION
**Pull request by ShanniLi**
r @Raynos @kriskowal

Second PR for #639 simplification.

### CODE REVIEW

File: node/connection_base.js [hunk 1]
```diff
@@ -216,6 +216,13 @@ TChannelConnectionBase.prototype.onReqError = function onReqError(req, err) {
 
 TChannelConnectionBase.prototype.runHandler = function runHandler(req) {
     var self = this;
+    self.channel.inboundCallsRecvdStat.increment(1, {
+        'calling-service': req.headers.cn,
+        'service': req.serviceName,
+        'endpoint': String(req.arg1)
+    });
+
+    req.latencyStart = self.timers.now();
```
Comment (Raynos) [inline node/connection_base.js]: no; this should be on the in_request.js itself as `self.start` field.

File: node/connection_base.js [hunk 2]
```diff
@@ -230,6 +237,13 @@ TChannelConnectionBase.prototype.buildResponse = function buildResponse(req, opt
             state: req.res.state
         }));
     }
+    options = extend({
```
Comment (Raynos) [inline node/connection_base.js]: move this stuff into `buildOutResponse` instead

Also just pass `inreq` to outResponse as a field if its not already tehre

File: node/out_response.js [hunk 3]
```diff
@@ -34,8 +34,14 @@ function TChannelOutResponse(id, options) {
     self.spanEvent = self.defineEvent('span');
     self.finishEvent = self.defineEvent('finish');
 
+    self.channel = options.channel;
+    self.callingService = options.callingService;
+    self.endpoint = options.endpoint;
     self.logger = options.logger;
+    self.latencyStart = options.latencyStart;
```
Comment (Raynos) [inline node/out_response.js]: this class already has `.start` and `.end`; use that.

File: node/out_response.js [hunk 4]
```diff
@@ -34,8 +34,14 @@ function TChannelOutResponse(id, options) {
     self.spanEvent = self.defineEvent('span');
     self.finishEvent = self.defineEvent('finish');
 
+    self.channel = options.channel;
+    self.callingService = options.callingService;
```
Comment (Raynos) [inline node/out_response.js]: use an `self.inreq` field instead of `callingService`, `endpoint`, `serviceName`

File: node/out_response.js [hunk 5]
```diff
@@ -89,6 +95,21 @@ TChannelOutResponse.prototype._sendError = function _sendError(codeString, messa
     });
 };
 
+TChannelOutResponse.prototype._emitLatency = function _emitLatency() {
```
Comment (Raynos) [inline node/out_response.js]: no need for underscore in method name.

File: node/out_response.js [hunk 6]
```diff
@@ -114,6 +135,15 @@ TChannelOutResponse.prototype.sendParts = function sendParts(parts, isLast) {
...
             break;
+        default:
+            self.channel.logger.error('TChannelOutResponse is in a wrong state', {
+                state: self.state
+            });
+            break;
+    }
+
+    if (isLast) {
+        self._emitLatency();
```
Comment (Raynos) [inline node/out_response.js]: not here. just call this in `onFinish()` instead.

File: node/out_response.js [hunk 7]
```diff
@@ -177,6 +207,7 @@ TChannelOutResponse.prototype.sendCallResponseContFrame = function sendCallRespo
 
 TChannelOutResponse.prototype.sendError = function sendError(codeString, message) {
     var self = this;
+    self._emitLatency();
```
Comment (Raynos) [inline node/out_response.js]: not here. just call this in `onFinish()`

**Comment by Raynos** [comment]
Finished review; we should shuffle some of this code around.

File: node/connection_base.js (hunk 2 above)
Comment (ShanniLi) [inline node/connection_base.js]: Is there any reason to move to buildOutResponse? Note that both self-connection and connection need it.

Comment (Raynos) [inline node/connection_base.js]: Ah then we can just change it to pass `inreq` in :)

**Comment by Raynos** [comment]
lgtm.

**Comment by kriskowal** [comment]
:ok: Thanks for addressing my feedback about assert.ifError, string concatenation, and default terms in stat names. :+1:

**Pull request merged**